api/v1alpha1/openapi.yaml

```diff
@@ -1622,19 +1622,54 @@ components:
         - metadata
         - spec
       description: Repository represents a git repository
-    RepositorySpec:
+    RepositoryHttpConfig:
       type: object
       properties:
-        repo:
-          type: string
-          description: 'The (possibly remote) repository URL to clone from'
         username:
           type: string
           description: 'The username for auth with HTTP transport'
         password:
           type: string
           description: 'The password for auth with HTTP transport'
           format: password
+        tls.crt:
+          type: string
+          description: 'Base64 encoded TLS cert data'
+          format: password
+        tls.key:
+          type: string
+          description: 'Base64 encoded TLS cert key'
+          format: password
+        ca.crt:
+          type: string
+          description: 'Base64 encoded root CA'
+        skipServerVerification:
+          type: boolean
+          description: 'Skip remote server verification'
+    RepositoryGitConfig:
+      type: object
+      properties:
+        sshPrivateKey:
+          type: string
+          description: 'The private SSH key'
+          format: password
+        privateKeyPassphrase:
+          type: string
+          description: 'The passphrase for sshPrivateKey'
+          format: password
+        skipServerVerification:
+          type: boolean
+          description: 'Skip remote server verification'
+    RepositorySpec:
+      type: object
+      properties:
```

Comment:
Let's use OneOf?

Reply:
Right, OneOf would make sense here.

Reply:
I've used `OneOf`. The schema looks like 
```yaml
GitHttpConfig:
      type: object
      properties:
        repo:
          type: string
          description: 'The (possibly remote) repository URL to clone from'
        username:
          type: string
          description: 'The username for auth with HTTP transport'
        password:
          type: string
          description: 'The password for auth with HTTP transport'
          format: password
        tls.crt:
          type: string
          description: 'Base64 encoded TLS cert data'
          format: password
        tls.key:
          type: string
          description: 'Base64 encoded TLS cert key'
          format: password
        ca.crt:
          type: string
          description: 'Base64 encoded root CA'
        skipServerVerification:
          type: boolean
          description: 'Skip remote server verification'
      required:
        - repo
    GitHttpConfig:
      type: object
      properties:
        repo:
          type: string
          description: 'The (possibly remote) repository URL to clone from'
        sshPrivateKey:
          type: string
          description: 'The private SSH key'
          format: password
        privateKeyPassphrase:
          type: string
          description: 'The passphrase for sshPrivateKey'
          format: password
        skipServerVerification:
          type: boolean
          description: 'Skip remote server verification'
      required:
        - repo
    RepositorySpec:
      type: object
      oneOf:
        - $ref: "#/components/schemas/GitHttpConfig"
        - $ref: "#/components/schemas/GitSshConfig"
```

However there were couple of issues.
 - the `GitHttpConfig` and `GitSshConfig` objects are clashing - if you try to create a repo you'll get a response error ```Error at "/spec": input matches more than one oneOf schemas```
  - I've added `additionalProperties: false` which solved the error.
  - oapi generator seems to be not handling the `additionalProperties: false` so I had to write our own `GetGitHttpConfig` / `GetGitSshConfig` parsers which are using strict json decoding (disallowing unknown fields)

Reply:
hmm, and now we can't handle the simplest repo example which has only `spec.repo` defined

Reply:
I see two options:
 go back to
```yaml
spec:
  repo: string
  httpConfig/sshConfig: {}

```

or keep using OneOf and add a discriminator field ie `type: 'ssh' | 'http'`

Reply:
Yes, we use a discriminator for device.Spec

Reply:
In the device.spec the discriminator is redundant and exposed these really clunky long internal names to users, so I anyway wanted to propose to re-discuss them. What is the issue with the PR's original proposal?

Reply:
The benefit to OneOf is that it's clear at the API level that only one of the properties should be populated at a given time, the API checks it, and the discriminator makes it easy for the code that deals with it to know what kind of resource it is (in this case, is it SSH or HTTP).

Reply:
In the latest changes I've moved the http/ssh config into a separate field `config`.
That way we make use of `oneOf` and `discriminator` is not needed.

WDYT ? @avishayt @fzdarsky

Reply:
I was OK with the discriminator.  I'll let @fzdarsky chime in.

Reply:
I believe usability is better without discriminator.